core/rawdb/freezer_batch.go

```diff
@@ -0,0 +1,220 @@
+		}
+		item = tb.curItem
+		count++
+	}
+
+	// Commit all table batches.
+	for _, tb := range batch.tables {
+		if err := tb.Commit(); err != nil {
+			return err
+		}
+	}
+
+	// Bump frozen block index.
+	atomic.AddUint64(&batch.freezer.frozen, uint64(count))
+	return nil
+}
+
+const (
+	freezerBatchBufferLimit = 2 * 1024 * 1024
+)
+
+// freezerTableBatch is a batch for a freezer table.
+type freezerTableBatch struct {
+	t *freezerTable
+
+	sb          *BufferedSnapWriter
+	encBuffer   writeBuffer
+	dataBuffer  []byte
+	indexBuffer []byte
+
+	curItem   uint64
+	headBytes uint32 // number of bytes in head file.
+}
+
+// newBatch creates a new batch for the freezer table.
+func (t *freezerTable) newBatch() *freezerTableBatch {
+	batch := &freezerTableBatch{t: t}
+	batch.Reset()
+	return batch
+}
+
+// Reset clears the batch for reuse.
+func (batch *freezerTableBatch) Reset() {
+	if !batch.t.noCompression {
+		batch.sb = new(BufferedSnapWriter)
+	} else {
+		batch.sb = nil
+	}
+	batch.encBuffer.Reset()
+	batch.dataBuffer = batch.dataBuffer[:0]
+	batch.indexBuffer = batch.indexBuffer[:0]
+	batch.curItem = atomic.LoadUint64(&batch.t.items)
+	batch.headBytes = batch.t.headBytes
+}
+
+// Append rlp-encodes and adds data at the end of the freezer table. The item number is a
+// precautionary parameter to ensure data correctness, but the table will reject already
+// existing data.
+func (batch *freezerTableBatch) Append(item uint64, data interface{}) error {
+	if item != batch.curItem {
+		return fmt.Errorf("appending unexpected item: want %d, have %d", batch.curItem, item)
+	}
+
+	// Encode the item.
+	batch.encBuffer.Reset()
+	if batch.sb != nil {
+		// RLP-encode
+		if err := rlp.Encode(batch.sb, data); err != nil {
+			return err
+		}
+		// Snappy-encode to our buf
+		if err := batch.sb.WriteTo(&batch.encBuffer); err != nil {
+			return err
+		}
+	} else {
+		if err := rlp.Encode(&batch.encBuffer, data); err != nil {
+			return err
+		}
+	}
+
+	return batch.appendItem(batch.encBuffer.data)
+}
+
+// AppendRaw injects a binary blob at the end of the freezer table. The item number is a
+// precautionary parameter to ensure data correctness, but the table will reject already
+// existing data.
+func (batch *freezerTableBatch) AppendRaw(item uint64, blob []byte) error {
+	if item != batch.curItem {
+		return fmt.Errorf("appending unexpected item: want %d, have %d", batch.curItem, item)
+	}
+
+	data := blob
+	if batch.sb != nil {
+		batch.encBuffer.Reset()
+		batch.sb.WriteDirectTo(&batch.encBuffer, blob)
+		data = batch.encBuffer.data
+	}
+	return batch.appendItem(data)
+}
+
+func (batch *freezerTableBatch) appendItem(data []byte) error {
+	itemSize := uint32(len(data))
+
+	// Check if item fits into current data file.
+	if batch.headBytes+itemSize > batch.t.maxFileSize {
+		// It doesn't fit, go to next file first.
+		if err := batch.commit(); err != nil {
+			return err
+		}
```

Comment:
If we want to break this invariant, and write until overflow instead of stopping before overflow, it should not cause any backward compatibility issues. Don't know if it makes any difference though?

Reply:
I've decided to keep this behavior for now.